File: 0063-unique-paths-ii/0063-unique-paths-ii.py
class Solution:
    # O(m*n) space
    def uniquePathsWithObstacles(self, obstacleGrid):
        if not obstacleGrid:
            return 
        r, c = len(obstacleGrid), len(obstacleGrid[0])
        dp = [[0 for _ in range(c)] for _ in range(r)]
        dp[0][0] = 1 - obstacleGrid[0][0]
        for i in range(1, r):
            dp[i][0] = dp[i-1][0] * (1 - obstacleGrid[i][0])
        for i in range(1, c):
            dp[0][i] = dp[0][i-1] * (1 - obstacleGrid[0][i])
        for i in range(1, r):
            for j in range(1, c):
                dp[i][j] = (dp[i][j-1] + dp[i-1][j]) * (1 - obstacleGrid[i][j])
        return dp[-1][-1]

#     # O(n) space
#     def uniquePathsWithObstacles2(self, obstacleGrid):
#         if not obstacleGrid:
#             return 
#         r, c = len(obstacleGrid), len(obstacleGrid[0])
#         cur = [0] * c
#         cur[0] = 1 - obstacleGrid[0][0]
#         for i in xrange(1, c):
#             cur[i] = cur[i-1] * (1 - obstacleGrid[0][i])
#         for i in xrange(1, r):
#             cur[0] *= (1 - obstacleGrid[i][0])
#             for j in xrange(1, c):
#                 cur[j] = (cur[j-1] + cur[j]) * (1 - obstacleGrid[i][j])
#         return cur[-1]

#     # in place
#     def uniquePathsWithObstacles(self, obstacleGrid):
#         if not obstacleGrid:
#             return 
#         r, c = len(obstacleGrid), len(obstacleGrid[0])
#         obstacleGrid[0][0] = 1 - obstacleGrid[0][0]
#         for i in xrange(1, r):
#             obstacleGrid[i][0] = obstacleGrid[i-1][0] * (1 - obstacleGrid[i][0])
#         for i in xrange(1, c):
#             obstacleGrid[0][i] = obstacleGrid[0][i-1] * (1 - obstacleGrid[0][i])
#         for i in xrange(1, r):
#             for j in xrange(1, c):
#                 obstacleGrid[i][j] = (obstacleGrid[i-1][j] + obstacleGrid[i][j-1]) * (1 - obstacleGrid[i][j])
#         return obstacleGrid[-1][-1]        

File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& arr) {
        int m=arr.size(),n=arr[0].size();
        if(arr[0][0]==1)return 0;
        vector<vector<int>>dp(m,vector<int>(n,0));
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(arr[i][j]==1)dp[i][j]=0;
            }
        }
         if(arr[0][0]==0)dp[0][0]=1;
        for(int j=1;j<n;j++){
            if(dp[0][j-1]==1 and arr[0][j]!=1)dp[0][j]=1;
            else dp[0][j]=0;
        }
        for(int i=1;i<m;i++){
            if(dp[i-1][0]==1 and arr[i][0]!=1)dp[i][0]=1;
            else dp[i][0]=0;
        }
        int ans=0;
        
        for(int i=1;i<m;i++){
            for(int j=1;j<n;j++){
                if(arr[i][j]==0){
                    dp[i][j]=dp[i-1][j]+dp[i][j-1];
                }
            }
        }
        return dp[m-1][n-1];
    }
};